fem/dfem/util.hpp

```diff
@@ -0,0 +1,2205 @@
+constexpr void for_constexpr_with_arg(lambda&& f, arg_t&& arg,
+                                      std::integer_sequence<std::size_t, i, Is...>)
+{
+   f(std::integral_constant<std::size_t, i> {}, get<i>(arg));
+   for_constexpr_with_arg(f, std::forward<arg_t>(arg),
+                          std::integer_sequence<std::size_t, Is...> {});
+}
+
+template <typename lambda, typename arg_t>
+constexpr void for_constexpr_with_arg(lambda&& f, arg_t&& arg)
+{
+   using indices =
+      std::make_index_sequence<tuple_size<std::remove_reference_t<arg_t>>::value>;
+   for_constexpr_with_arg(std::forward<lambda>(f), std::forward<arg_t>(arg),
+                          indices{});
+}
+
+template <typename... input_ts, std::size_t... Is>
+auto make_dependency_map_impl(
+   tuple<input_ts...> inputs,
+   std::index_sequence<Is...>)
+{
+   auto make_dependency_array = [&](auto i)
+   {
+      return std::array<bool, sizeof...(input_ts)>
+      {
+         (get<i>(inputs).GetFieldId() == get<Is>(inputs).GetFieldId())...
+      };
+   };
+
+   std::unordered_map<int, std::array<bool, sizeof...(input_ts)>> map;
+   for_constexpr<sizeof...(input_ts)>([&](auto i)
+   {
+      map[get<i>(inputs).GetFieldId()] =
+         make_dependency_array(std::integral_constant<std::size_t, i> {});
+   });
+
+   return map;
+}
+
+// @brief Create a dependency map from a tuple of inputs.
+//
+// @param inputs a tuple of objects derived from FieldOperator.
+// @returns an unordered_map where the keys are the field IDs and the values
+// are arrays of booleans indicating which inputs depend on each field ID.
+template <typename... input_ts>
+auto make_dependency_map(tuple<input_ts...> inputs)
+{
+   return make_dependency_map_impl(inputs, std::index_sequence_for<input_ts...> {});
+}
+
+// @brief Get the type name of a template parameter T.
+//
+// Convenient helper function for debugging.
+// Usage example
+// ```c++
+// mfem::out << get_type_name<int>() << std::endl;
+// ```
+// prints "int".
+template <typename T>
+constexpr auto get_type_name() -> std::string_view
+{
+#if defined(__clang__)
+   constexpr auto prefix = std::string_view {"[T = "};
+   constexpr auto suffix = "]";
+   constexpr auto function = std::string_view{__PRETTY_FUNCTION__};
+#elif defined(__GNUC__)
+   constexpr auto prefix = std::string_view {"with T = "};
+   constexpr auto suffix = "; ";
+   constexpr auto function = std::string_view{__PRETTY_FUNCTION__};
+#elif defined(_MSC_VER)
+   constexpr auto prefix = std::string_view {"get_type_name<"};
+   constexpr auto suffix = ">(void)";
+   constexpr auto function = std::string_view{__FUNCSIG__};
+#else
+#error Unsupported compiler
+#endif
+
+   const auto start = function.find(prefix) + prefix.size();
+   const auto end = function.find(suffix);
+   const auto size = end - start;
+
+   return function.substr(start, size);
+}
+
+template <typename Tuple, std::size_t... Is>
+void print_tuple_impl(const Tuple& t, std::index_sequence<Is...>)
```

Comment:
```suggestion
inline void print_tuple_impl(const Tuple& t, std::index_sequence<Is...>)
```
as well as for other functions in this file, to avoid potential `duplicate symbol`.